NBitcoin/Protocol/NetworkAddress.cs

```diff
@@ -5,113 +5,421 @@
+				}
+				else if (value is DnsEndPoint dnsEndPoint)
+				{
+					SetSpecial(dnsEndPoint.Host);
+					port = (ushort)dnsEndPoint.Port;
+				}
+				else
+				{
+					throw new ArgumentException($"Not supported {nameof(EndPoint)} type {value.GetType().FullName}.");
+				}
 			}
+		} 
+
+		protected void SetIp(IPAddress ip)
+		{
+			SetLegacyIpv6(ip.MapToIPv6().GetAddressBytes());
 		}
 
-		public TimeSpan Ago
+		public void SetLegacyIpv6(byte[] ipv6)
 		{
-			get
+			if (ipv6.Length != ADDR_IPV6_SIZE)
 			{
-				return DateTimeOffset.UtcNow - Time;
+				throw new ArgumentException("The received value is not a valid IPv6 byte array.", nameof(ipv6));
 			}
-			set
+
+			var skip = 0;
+			if (HasPrefix(ipv6, IPV4_IN_IPV6_PREFIX))
 			{
-				Time = DateTimeOffset.UtcNow - value;
+				network = (byte)Network.IPv4;
+				skip = IPV4_IN_IPV6_PREFIX.Length;
 			}
-		}
+			else if(HasPrefix(ipv6, TORV2_IN_IPV6_PREFIX))
+			{
+				// TORv2-in-IPv6
+				network = (byte)Network.Onion;
+				skip = TORV2_IN_IPV6_PREFIX.Length;
+			} 
+			else
+			{
+				// IPv6
+				network = (byte)Network.IPv6;
+			}	
 
-		public void Adjust()
-		{
-			var nNow = Utils.DateTimeToUnixTime(DateTimeOffset.UtcNow);
-			if (ntime <= 100000000 || ntime > nNow + 10 * 60)
-				ntime = nNow - 5 * 24 * 60 * 60;
+			addr = ipv6.Skip(skip).ToArray();
 		}
 
-		public IPEndPoint Endpoint
+		private bool SetTor(string address)
 		{
-			get
+			if (!address.EndsWith(".onion"))
 			{
-				return new IPEndPoint(new IPAddress(ip), port);
+				return false;
 			}
-			set
+
+			byte[] bytes;
+			try
+			{
+				bytes = Encoders.Base32.DecodeData(address.Substring(0, address.Length - ".onion".Length));
+			}
+			catch(FormatException)
 			{
-				port = (ushort)value.Port;
-				var ipBytes = value.Address.GetAddressBytes();
-				if (ipBytes.Length == 16)
+				return false;
+			}
+
+			if (bytes.Length == ADDR_TORV2_SIZE)
+			{
+				network = (byte)Network.Onion;
+				addr = bytes;
+				return true;
+			}
+			else if(bytes.Length == ADDR_TORV3_SIZE  + TORV3_ADDR_CHECKSUM_LEN + TORV3_ADDR_VERSION_LEN)
+			{
+				var version = bytes[bytes.Length - 1];
+
+				if (version != 3)
 				{
-					ip = ipBytes;
+					return false;
 				}
-				else if (ipBytes.Length == 4)
+
+				var pubkey = bytes.SafeSubarray( 0, ADDR_TORV3_SIZE);
+				var chksum = bytes.SafeSubarray(ADDR_TORV3_SIZE,  TORV3_ADDR_CHECKSUM_LEN);
+
+				var calculatedChecksum = CalculateChecksum(pubkey);
+
+				if (!Utils.ArrayEqual(chksum, calculatedChecksum) )
 				{
-					//Convert to ipv4 mapped to ipv6
-					//In these addresses, the first 80 bits are zero, the next 16 bits are one, and the remaining 32 bits are the IPv4 address
-					ip = new byte[16];
-					Array.Copy(ipBytes, 0, ip, 12, 4);
-					Array.Copy(new byte[] { 0xFF, 0xFF }, 0, ip, 10, 2);
+					return false;
 				}
-				else
-					throw new NotSupportedException("Invalid IP address type");
+
+				network = (byte)Network.Onion;
+				addr = pubkey;

```

Comment:
This is a bit different from what Bitcoin does. This replaces the original array with a new one. Bitcoin uses a byte vector whose capacity is increased as needed.

I mention this just because I got confused by the initialization `private byte[] addr = new byte[ADDR_IPV6_SIZE];`.

Reply:
Yes, we don't really need to initialized it to 16 bytes but it has to have a initial value.